core-tests/shared/src/test/scala/zio/stm/TArraySpec.scala

```diff
@@ -0,0 +1,204 @@
+              lengths      <- lengthsFiber.join
+              firstAndLast <- lengths.array(0).get.zip(lengths.array(N - 1).get).commit
+            } yield assert(firstAndLast, equalTo((19, 19)) || equalTo((3, 3)))
+          }
+        ),
+        suite("mapM")(
+          testM("creates new array atomically") {
+            for {
+              tArray       <- makeTArray(N)("thisStringLengthIs20")
+              lengthsFiber <- tArray.mapM(a => STM.succeed(a.length)).commit.fork
+              _            <- STM.foreach(0 until N)(idx => tArray.array(idx).set("abc")).commit
+              lengths      <- lengthsFiber.join
+              first        <- lengths.array(0).get.commit
+              last         <- lengths.array(N - 1).get.commit
+            } yield assert((first, last), equalTo((20, 20)) || equalTo((3, 3)))
+          },
+          testM("returns effect failure") {
+            for {
+              tArray <- makeTArray(N)("abc")
+              _      <- tArray.array(N / 2).update(_ => "").commit
+              result <- tArray.mapM(a => if (a.isEmpty) STM.fail(boom) else STM.succeed(())).commit.either
+            } yield assert(result, isLeft(equalTo(boom)))
+          }
+        ),
+        suite("transform")(
+          testM("updates values atomically") {
+            for {
+              tArray         <- makeTArray(N)("a")
+              transformFiber <- tArray.transform(_ + "+b").commit.fork
+              _              <- STM.foreach(0 until N)(idx => tArray.array(idx).update(_ + "+c")).commit
+              _              <- transformFiber.join
+              first          <- tArray.array(0).get.commit
+              last           <- tArray.array(N - 1).get.commit
+            } yield assert((first, last), equalTo(("a+b+c", "a+b+c")) || equalTo(("a+c+b", "a+c+b")))
+          }
+        ),
+        suite("transformM")(
+          testM("updates values atomically") {
+            for {
+              tArray         <- makeTArray(N)("a")
+              transformFiber <- tArray.transformM(a => STM.succeed(a + "+b")).commit.fork
+              _              <- STM.foreach(0 until N)(idx => tArray.array(idx).update(_ + "+c")).commit
+              _              <- transformFiber.join
+              first          <- tArray.array(0).get.commit
+              last           <- tArray.array(N - 1).get.commit
+            } yield assert((first, last), equalTo(("a+b+c", "a+b+c")) || equalTo(("a+c+b", "a+c+b")))
+          },
+          testM("updates all or nothing") {
+            for {
+              tArray <- makeTArray(N)(0)
+              _      <- tArray.array(N / 2).update(_ => 1).commit
+              result <- tArray.transformM(a => if (a == 0) STM.succeed(42) else STM.fail(boom)).commit.either
+              first  <- tArray.array(0).get.commit
+            } yield assert(result.left.map(r => (first, r)), isLeft(equalTo((0, boom))))
```

Comment:
This is not great too. Not sure how to improve that. The original was just
```
(first, result) === (0, Left(boom))
```

Reply:
Looks good for me!
Alternatively you could manually annotate type `assert((first, result), equalTo((0, Right(()): Either[Exception, Unit])))` but I don't find it superior in any way.
Maybe there is some way of improving type inference for `equalTo`, but perhaps out of scope here.